File: FAO_case_study/algorithms/MPFTM/task_migration_based_pon.py
import math
import networkx as nx
from typing import Dict, List, Set
from input.group import Group
from input.robot import Robot
from input.task import Task
from input.potential_field import PotentialField
from input.migration_record import MigrationRecord
from .ini_context_load_i import IniContextLoadI
from .calculate_pon_field import CalculatePonField


class TaskMigrationBasedPon:
    def __init__(self, id_to_groups: Dict[int, Group], id_to_robots: Dict[int, Robot],
                 arc_graph: nx.Graph, group_id_to_pfield: Dict[int, PotentialField],
                 robot_id_to_pfield: Dict[int, PotentialField], shortest_path_dict: Dict,
                 id_to_i: Dict[int, float], a: float, b: float):
        self.id_to_groups = id_to_groups
        self.id_to_robots = id_to_robots
        self.arc_graph = arc_graph
        self.group_id_to_pfield = group_id_to_pfield
        self.robot_id_to_pfield = robot_id_to_pfield
        self.shortest_path_dict = shortest_path_dict
        self.a = a
        self.b = b
        self.id_to_i = id_to_i
        self.records = []

    def run(self) -> List[MigrationRecord]:
        """Execute task migration based on potential field."""
        self._inter_task_migration()
        return self.records

    def _inter_task_migration(self):
        """Inter-group task migration."""
        f_groups = set()
        for robot_id in self.id_to_robots.keys():
            robot = self.id_to_robots[robot_id]
            if robot.fault_a == 1:
                f_groups.add(robot.group_id)

        average_pe_n = self._get_average_pe_n()

        for fgroup_id in f_groups:
            s_group = self.id_to_groups[fgroup_id]
            robot_id_in_group = s_group.robot_id_in_group

            for robot_id in robot_id_in_group:
                robot = self.id_to_robots[robot_id]
                if robot.fault_a == 1:
                    # pf represents potential field of faulty network layer
                    tnf = list(robot.tasks_list)
                    pf = self.group_id_to_pfield[fgroup_id]
                    p_fg = pf.pegra + pf.perep

                    if p_fg > average_pe_n:
                        # Need inter-layer task migration
                        t_group_id = self._find_min_pn()
                        for task in tnf:
                            pt = self.group_id_to_pfield[t_group_id]
                            p_tg = pt.pegra + pt.perep
                            if p_tg < average_pe_n:
                                self._execute_migration(robot, self.id_to_groups[t_group_id].leader, task)

            # Execute intra-group task migration
            self._intra_task_migration(fgroup_id)

    def _intra_task_migration(self, group_id: int):
        """Intra-group task migration (includes recursive algorithm)."""
        f_robots = []
        group = self.id_to_groups[group_id]

        for robot_id in group.robot_id_in_group:
            robot = self.id_to_robots[robot_id]
            if robot.fault_a == 1:
                f_robots.append(robot)

        leader = group.leader

        # Migrate tasks from faulty robots to prevent cascading failures
        for f_robot in f_robots:
            tasks_list = f_robot.tasks_list
            while len(tasks_list) > 0:
                migrated_task = tasks_list[0]
                migrated_robot = self._find_migrated_robot(f_robot)
                self._execute_migration(f_robot, migrated_robot, migrated_task)
                tasks_list = f_robot.tasks_list

                # Update tasksList
                self._migration_for_robot(migrated_robot)

        self._migration_for_robot(leader)

    def _migration_for_robot(self, robot: Robot):
        """Execute migration for a specific robot."""
        robot_id = robot.robot_id
        neighbors = list(self.arc_graph.neighbors(robot_id))
        domain_id = []

        for neighbor_id in neighbors:
            domain_id.append(neighbor_id)

        domain_id.sort(key=lambda x: self._get_comparator_value(robot, x))

        if not domain_id:
            return

        migrated_id = domain_id[0]

        po_r = self.robot_id_to_pfield[robot_id]
        por_value = po_r.pegra + po_r.perep

        po_m = self.robot_id_to_pfield[migrated_id]
        pom_value = po_m.pegra + po_m.perep

        tasks_list = robot.tasks_list
        migrated_task = self._find_max_task(tasks_list)

        if not self.arc_graph.has_edge(robot_id, migrated_id):
            return

        c = self.arc_graph[robot_id][migrated_id]['weight']

        while ((por_value - pom_value) / c) > 0.02:
            robot_migrated = self.id_to_robots[migrated_id]
            self._execute_migration(robot, robot_migrated, migrated_task)
            self._migration_for_robot(robot_migrated)

            # Continue recursion from ai node
            domain_id.sort(key=lambda x: self._get_comparator_value(robot, x))
            if not domain_id:
                break

            migrated_id = domain_id[0]
            po_r = self.robot_id_to_pfield[robot_id]
            por_value = po_r.pegra + po_r.perep
            po_m = self.robot_id_to_pfield[migrated_id]
            pom_value = po_m.pegra + po_m.perep

    def _get_comparator_value(self, robot: Robot, neighbor_id: int) -> float:
        """Calculate comparison value for sorting neighbors."""
        po1 = self.robot_id_to_pfield[neighbor_id]
        po1_value = po1.pegra + po1.perep

        robot_id = robot.robot_id
        po_m = self.robot_id_to_pfield[robot_id]
        po_m_value = po_m.pegra + po_m.perep

        if not self.arc_graph.has_edge(robot_id, neighbor_id):
            return float('inf')

        cij1 = self.arc_graph[robot_id][neighbor_id]['weight']

        return -((po1_value - po_m_value) / cij1)  # Negate for descending order

    def _find_max_task(self, tasks_list: List[Task]) -> Task:
        """Find task with maximum size."""
        if not tasks_list or len(tasks_list) < 1:
            return None
        return max(tasks_list, key=lambda t: t.size)

    def _find_migrated_robot(self, f_robot: Robot) -> Robot:
        """Find robot to migrate tasks to."""
        migrated_robot = Robot()
        neighbors = list(self.arc_graph.neighbors(f_robot.robot_id))
        min_value = float('inf')

        for neighbor_id in neighbors:
            target_robot = self.id_to_robots[neighbor_id]

            target_p = self.robot_id_to_pfield[target_robot.robot_id]
            weight = self.arc_graph[f_robot.robot_id][neighbor_id]['weight']
            v = (target_p.pegra + target_p.perep) * weight

            if v < min_value:
                migrated_robot = target_robot
                min_value = v

        return migrated_robot

    def _find_min_pn(self) -> int:
        """Find group with minimum potential field."""
        min_value = float('inf')
        return_id = -1

        for group_id in self.group_id_to_pfield.keys():
            p = self.group_id_to_pfield[group_id]
            p_value = p.perep + p.pegra
            if min_value > p_value:
                min_value = p_value
                return_id = group_id

        return return_id

    def _get_average_pe_n(self) -> float:
        """Calculate average potential field of network layer."""
        pe_n_sum = 0.0
        for group_id in self.group_id_to_pfield.keys():
            pe_n = self.group_id_to_pfield[group_id]
            pe_n_sum += pe_n.pegra + pe_n.perep

        return pe_n_sum / len(self.group_id_to_pfield) if self.group_id_to_pfield else 0.0

    def _execute_migration(self, robot: Robot, robot_migrated: Robot, migration_task: Task):
        """Execute task migration from one robot to another."""
        if robot is None or robot_migrated is None or migration_task is None:
            return

        # Update inter-layer load and task list if robots are in different groups
        if robot.group_id != robot_migrated.group_id:
            self._update_inter(robot, robot_migrated, migration_task)

        # Update intra-layer load and task list
        self._update_intra(robot, robot_migrated, migration_task)

        record = MigrationRecord()
        record.from_robot = robot.robot_id
        record.to_robot = robot_migrated.robot_id
        self.records.append(record)

        # Re-initialize contextual load
        ini_context = IniContextLoadI(self.id_to_groups, self.id_to_robots, self.arc_graph,
                                     self.shortest_path_dict, self.id_to_i, self.a, self.b)
        ini_context.run()

        # Update potential field
        calculate_pon_field = CalculatePonField(self.id_to_groups, self.id_to_robots,
                                               self.arc_graph, self.id_to_i,
                                               self.shortest_path_dict, self.a, self.b)

        if robot.group_id != robot_migrated.group_id:
            # Update network layer potential field
            self.group_id_to_pfield = calculate_pon_field.calculate_inter_p()

        # Update node potential field
        self.robot_id_to_pfield = calculate_pon_field.calculate_intra_p()

    def _update_inter(self, robot: Robot, robot_migrated: Robot, migration_task: Task):
        """Update inter-group load and task list."""
        if robot is None or robot_migrated is None or migration_task is None:
            return

        group_id = robot.group_id
        group = self.id_to_groups[group_id]
        robot_migrated_group_id = robot_migrated.group_id
        migrated_group = self.id_to_groups[robot_migrated_group_id]

        group.group_load = group.group_load - migration_task.size
        migrated_group.group_load = migrated_group.group_load + migration_task.size

    def _update_intra(self, robot: Robot, robot_migrated: Robot, migration_task: Task):
        """Update intra-group load and task list after migration."""
        tasks_list = robot.tasks_list
        if migration_task in tasks_list:
            tasks_list.remove(migration_task)
        robot.load = robot.load - migration_task.size
        robot.tasks_list = tasks_list

        robot_migrated_task_list = robot_migrated.tasks_list
        if robot_migrated_task_list is None:
            robot_migrated_task_list = []

        robot_migrated_task_list.append(migration_task)
        robot_migrated.load = robot_migrated.load + migration_task.size
        robot_migrated.tasks_list = robot_migrated_task_list

File: cascadingFailuresTaskMigration/src/main/java/MPFTM/TaskMigrationBasedPon.java
package MPFTM;

import input.*;
import org.jgrapht.alg.interfaces.ShortestPathAlgorithm;
import org.jgrapht.graph.DefaultUndirectedWeightedGraph;
import org.jgrapht.graph.DefaultWeightedEdge;

import java.util.*;

public class TaskMigrationBasedPon {
    private DefaultUndirectedWeightedGraph<Integer, DefaultWeightedEdge> arcGraph;
    private HashMap<Integer, Group> idToGroups;
    private HashMap<Integer,Robot> idToRobots;
    private ShortestPathAlgorithm<Integer, DefaultWeightedEdge> shortestPath ;
    private HashMap<Integer,PotentialField> groupIdToPfield;
    private HashMap<Integer,PotentialField> robotIdToPfield;
    private List<MigrationRecord> records;
    Double a;
    Double b;
    private HashMap<Integer,Double> idToI;
    public TaskMigrationBasedPon(HashMap<Integer, Group> idToGroups, HashMap<Integer, Robot> idToRobots,
                                 DefaultUndirectedWeightedGraph<Integer, DefaultWeightedEdge> arcGraph, HashMap<Integer, PotentialField> groupIdToPfield, HashMap<Integer, PotentialField> robotIdToPfield
            ,ShortestPathAlgorithm<Integer, DefaultWeightedEdge> shortestPath,HashMap<Integer,Double> idToI,Double a,Double b) {
        this.idToGroups = idToGroups;
        this.idToRobots = idToRobots;
        this.groupIdToPfield = groupIdToPfield;
        this.robotIdToPfield = robotIdToPfield;
        this.arcGraph = arcGraph;
        this.a = a;
        this.b = b;
        this.idToI = idToI;
        this.shortestPath = shortestPath ;
        records = new ArrayList<>();
    }

    public List<MigrationRecord> run() {
        InterTaskMigration();
        return records;
    }

    private void InterTaskMigration() {
        Set<Integer> Fgroups = new HashSet<>();
        for (Integer id : idToRobots.keySet()) {
            Robot robot = idToRobots.get(id);
            if (robot.getFaultA()==1) {
                Fgroups.add(robot.getGroupId());
            }
        }

        Double averagePeN = getAveragePeN();

        for (Integer fgroupId : Fgroups) {
            Group sGroup = idToGroups.get(fgroupId);
            Set<Integer> robotIdInGroup = sGroup.getRobotIdInGroup();
            for (Integer robotId : robotIdInGroup) {
                Robot robot = idToRobots.get(robotId);
                if (robot.getFaultA()==1) {
                    //pf表示发生故障的网络层的势场情况
                    List<Task> tnf = new ArrayList<>(robot.getTasksList());
                    PotentialField pf = groupIdToPfield.get(fgroupId);
                    Double pFg = pf.getPegra()+pf.getPerep();
                    if (pFg >averagePeN) {
                        //需要进行网络层间的任务迁移
                        int tGroupId = findMinPn();
                        for (Task task : tnf) {
                            PotentialField pt = groupIdToPfield.get(tGroupId);
                            Double pTg = pt.getPegra()+pt.getPerep();
                            if (pTg <averagePeN) {
                                excuteMigration(robot,idToGroups.get(tGroupId).getLeader(),task);
                            }
                        }
                    }

                }
            }
            //执行网络层内的任务迁移，对该group内部的任务进行迁移
            IntraTaskMigration(fgroupId);
        }
    }

    private void IntraTaskMigration(int groupId) {
        //执行网络层内的任务迁移，对该group内部的任务进行迁移（包含递归算法）
        List<Robot> fRobots = new ArrayList<>();
        Group group = idToGroups.get(groupId);
        for (Integer robotId : group.getRobotIdInGroup()) {
            Robot robot = idToRobots.get(robotId);
            if (robot.getFaultA()==1) {
                fRobots.add(robot);
            }
        }
        Robot leader = group.getLeader();
        //把需要的被迁移的任务 迁移给边缘的节点上，阻碍级联故障的发生（不设置这种情况了）
        for (Robot fRobot : fRobots) {
            List<Task> tasksList = fRobot.getTasksList();
            while (tasksList.size()>0) {
                Task migratedTask = tasksList.get(0);
                Robot migratedRobot = findMigratedRobot(fRobot);
                excuteMigration(fRobot,migratedRobot,migratedTask);
                tasksList = fRobot.getTasksList();

                //更新tasksList
                MigrationforRobot(migratedRobot);
            }
        }
        MigrationforRobot(leader);

    }

    private void MigrationforRobot(Robot robot) {
        int robotId = robot.getRobotId();
        Set<DefaultWeightedEdge> defaultWeightedEdges = arcGraph.edgesOf(robotId);
        List<Integer> domainId = new ArrayList<>();
        for (DefaultWeightedEdge defaultWeightedEdge : defaultWeightedEdges) {
            Integer edgeSource = arcGraph.getEdgeSource(defaultWeightedEdge);
            if (edgeSource!=robotId) {
                domainId.add(edgeSource);
            } else {
                Integer edgeTarget = arcGraph.getEdgeTarget(defaultWeightedEdge);
                domainId.add(edgeTarget);
            }
        }
        domainId.sort(getComparator(robot));

        Integer migratedId = domainId.get(0);

        PotentialField poR = robotIdToPfield.get(robotId);

        Double porValue = poR.getPegra()+poR.getPerep();

        PotentialField poM = robotIdToPfield.get(migratedId);

        Double pomValue = poM.getPegra()+poM.getPerep();

        List<Task> tasksList = robot.getTasksList();

        Task migratedTask = findMaxTask(tasksList);

        double c = arcGraph.getEdgeWeight(arcGraph.getEdge(robotId, migratedId));
        while (((porValue-pomValue)/c)>0.02) {
            Robot robotMigrated = idToRobots.get(migratedId);
            excuteMigration(robot, robotMigrated,migratedTask);
            MigrationforRobot(robotMigrated);
            //执行递归

            //从ai节点继续执行递归
            domainId.sort(getComparator(robot));
            migratedId = domainId.get(0);
            poR = robotIdToPfield.get(robotId);
            porValue = poR.getPegra()+poR.getPerep();
            poM = robotIdToPfield.get(migratedId);
            pomValue = poM.getPegra()+poM.getPerep();
        }

    }

    private Task findMaxTask(List<Task> tasksList) {
        if (tasksList==null||tasksList.size()<1) {
            return null;
        }
        Task returnTask = tasksList.get(0);
        for (Task task : tasksList) {
            if (task.getSize()>returnTask.getSize()) {
                returnTask = task;
            }
        }
        return returnTask;
    }

    private Comparator<Integer> getComparator(Robot robot) {
        return new Comparator<Integer>() {
                          @Override
                          public int compare(Integer o1, Integer o2) {
                              PotentialField po1 = robotIdToPfield.get(o1);
                              Double po1Value = po1.getPegra()+po1.getPerep();

                              PotentialField po2 = robotIdToPfield.get(o2);
                              Double po2Value = po2.getPegra()+po2.getPerep();

                              int robotId = robot.getRobotId();
                              PotentialField poM = robotIdToPfield.get(robotId);
                              Double poMValue = poM.getPegra()+poM.getPerep();

                              Double cij1 = arcGraph.getEdgeWeight(arcGraph.getEdge(robotId,o1));

                              Double cij2 = arcGraph.getEdgeWeight(arcGraph.getEdge(robotId,o2));

                              return (int) Math.ceil((po2Value-poMValue)/cij2 - (po1Value-poMValue)/cij1 );
                          }
                      };
    }

    private Robot findMigratedRobot(Robot fRobot) {
        Robot MigratedRobot = new Robot();
        Set<DefaultWeightedEdge> defaultWeightedEdges = arcGraph.edgesOf(fRobot.getRobotId());
        Double minValue = Double.MAX_VALUE;
        for (DefaultWeightedEdge defaultWeightedEdge : defaultWeightedEdges) {
            Robot targetRobot = idToRobots.get(arcGraph.getEdgeTarget(defaultWeightedEdge));
            if (targetRobot.getRobotId()==fRobot.getRobotId()) {
                targetRobot = idToRobots.get(arcGraph.getEdgeSource(defaultWeightedEdge));
            }
            PotentialField targetP = robotIdToPfield.get(targetRobot.getRobotId());
            Double v = (targetP.getPegra()+targetP.getPerep())*arcGraph.getEdgeWeight(defaultWeightedEdge);
            if (v <minValue) {
                MigratedRobot = targetRobot;
                minValue = v;
            }
        }
        return MigratedRobot;
    }

    private int findMinPn() {
        Double minValue = Double.MAX_VALUE;
        int returnId = -1;
        for (Integer groupId : groupIdToPfield.keySet()) {
            PotentialField p = groupIdToPfield.get(groupId);
            double pValue = p.getPerep() + p.getPegra();
            if (minValue> pValue) {
                minValue = pValue;
                returnId = groupId;
            }
        }
        return returnId;
    }


    private Double getAveragePeN() {
        Double peNsum = 0.0;
        for (Integer groupId : groupIdToPfield.keySet()) {
            PotentialField peN = groupIdToPfield.get(groupId);
            peNsum+= peN.getPegra();
            peNsum+=peN.getPerep();
        }
       return peNsum/groupIdToPfield.size();
    }

    private void excuteMigration(Robot robot, Robot robotMigrated,Task migrationTask) {
        if (robot == null) {
            return;
        }

        if (robotMigrated == null) {
            return;
        }
        if (migrationTask == null) {
            return;
        }

        //改变网络层间负载和任务列表
        if (robot.getGroupId()!=robotMigrated.getGroupId()) {
            //不在同一组才会更新网络层间的势场情况
            updateInter(robot, robotMigrated, migrationTask);
        }
        //改变网络层内负载和任务列表
        updateIntra(robot, robotMigrated, migrationTask);
        MigrationRecord record = new MigrationRecord();
        record.setFrom(robot.getRobotId());
        record.setTo(robotMigrated.getRobotId());
        records.add(record);
        //初始化（计算）上下文负载
        new IniContextLoadI(idToGroups,idToRobots,arcGraph, shortestPath,idToI,a, b).run();

        //更新势场情况
        CalculatePonField calculatePonField = new CalculatePonField(idToGroups, idToRobots, arcGraph,idToI,shortestPath,a,b);

        if (robot.getGroupId()!=robotMigrated.getGroupId()) {
            //更新网络层的势场
            groupIdToPfield = calculatePonField.calculateInterP();

        }

        //更新节点的势场
        robotIdToPfield = calculatePonField.calculateIntraP();

    }

    private void updateInter(Robot robot, Robot robotMigrated, Task migrationTask) {
        if (robot == null) {
            return;
        }

        if (robotMigrated == null) {
            return;
        }
        if (migrationTask == null) {
            return;
        }

        int groupId = robot.getGroupId();
        Group group = idToGroups.get(groupId);
        int robotMigratedGroupId = robotMigrated.getGroupId();
        Group migratedGroup = idToGroups.get(robotMigratedGroupId);

        group.setGroupLoad(group.getGroupLoad()-migrationTask.getSize());
        migratedGroup.setGroupLoad(migratedGroup.getGroupLoad()+migrationTask.getSize());


    }

    private void updateIntra(Robot robot, Robot robotMigrated, Task migrationTask) {
        //更新迁移后的任务负载
        List<Task> tasksList = robot.getTasksList();
        tasksList.remove(migrationTask);
        robot.setLoad(robot.getLoad()-migrationTask.getSize());
        robot.setTasksList(tasksList);


        List<Task> robotMigratedTaskList = robotMigrated.getTasksList();
        if (robotMigratedTaskList==null) {
            robotMigratedTaskList = new ArrayList<>();
        }

        robotMigratedTaskList.add(migrationTask);
        robotMigrated.setLoad(robotMigrated.getLoad()+migrationTask.getSize());
        robotMigrated.setTasksList(robotMigratedTaskList);

    }

}
